File: src/type/v2/subplot.ts
/**
 * Subplot（サブプロット）型定義
 *
 * メインプロットとサブプロットを管理し、物語の構造を表現する。
 *
 * @module
 */

// ========================================
// 基本型
// ========================================

/**
 * プロットタイプ
 * - main: メインプロット（物語の中心）
 * - subplot: サブプロット（メインを補完）
 * - parallel: 並行プロット（メインと独立して進行）
 * - background: 背景プロット（世界観を補強）
 */
export type PlotType = "main" | "subplot" | "parallel" | "background";

/**
 * プロットの重要度
 */
export type PlotImportance = "major" | "minor" | "supporting";

/**
 * 影響レベル
 */
export type InfluenceLevel = "low" | "medium" | "high";

/**
 * 影響方向
 * - forward: 元プロットから対象プロットへ影響
 * - backward: 対象プロットから元プロットへ影響
 * - mutual: 双方向の影響
 */
export type InfluenceDirection = "forward" | "backward" | "mutual";

// ========================================
// フォーカスキャラクター
// ========================================

/**
 * フォーカスキャラクター
 * プロット内でのキャラクターの役割を定義
 */
export type FocusCharacter = {
  /** キャラクターID */
  characterId: string;
  /** 重み（primary: 主役、secondary: 副次的） */
  weight: "primary" | "secondary";
};

// ========================================
// プロットビート
// ========================================

/**
 * プロットビート（起承転結の各ポイント）
 */
export type PlotBeat = {
  // 必須フィールド
  /** 一意なID */
  id: string;
  /** ビートタイトル */
  title: string;
  /** 短い概要 */
  summary: string;
  /** 関連チャプター */
  chapter: string;

  // 関連エンティティ
  /** 関連キャラクターIDリスト */
  characters: string[];
  /** 関連設定IDリスト */
  settings: string[];

  // オプション: 構造情報
  /** 構造位置（"setup", "rising", "climax", "falling", "resolution" など） */
  structurePosition?: string;
  /** カスタムステータス */
  customStatus?: string;

  // オプション: 因果関係（片方向のみ）
  /** このビートの原因となるビートID（前提条件） */
  preconditionBeatIds?: string[];

  // オプション: TimelineEventとのリンク
  /** 関連するTimelineEventのID */
  timelineEventId?: string;

  // オプション: 表示名
  /** 表示名のバリエーション */
  displayNames?: string[];
};

// ========================================
// プロット交差
// ========================================

/**
 * プロット交差
 * 異なるプロット間でビート同士がどのように影響し合うかを定義
 */
export type PlotIntersection = {
  // 必須: 交差元
  /** 交差元プロットID */
  sourcePlotId: string;
  /** 交差元ビートID */
  sourceBeatId: string;

  // 必須: 交差先
  /** 交差先プロットID */
  targetPlotId: string;
  /** 交差先ビートID */
  targetBeatId: string;

  // 交差の性質
  /** 交差の説明 */
  description: string;
  /** 影響方向 */
  influenceDirection: InfluenceDirection;
  /** 影響レベル */
  influenceLevel: InfluenceLevel;

  // オプション: 関連チャプター
  /** 交差が発生するチャプター */
  chapter?: string;
};

// ========================================
// 構造テンプレート
// ========================================

/**
 * 構造テンプレート参照
 */
export type StructureTemplate = {
  /** テンプレートID */
  id: string;
  /** テンプレート名 */
  name: string;
  /** 構造位置リスト */
  positions: string[];
};

// ========================================
// サブプロット詳細情報
// ========================================

/**
 * サブプロット詳細情報
 */
export type SubplotDetails = {
  /** 詳細説明 */
  description?: string | { file: string };
  /** このプロットの動機 */
  motivation?: string | { file: string };
  /** 結末の詳細 */
  resolution?: string | { file: string };
};

/**
 * LSP用検出ヒント
 */
export type SubplotDetectionHints = {
  /** よく使われるパターン */
  commonPatterns?: string[];
  /** 除外パターン */
  excludePatterns?: string[];
  /** 信頼度 */
  confidence?: number;
};

// ========================================
// サブプロット型（メイン）
// ========================================

/**
 * サブプロット型
 */
export type Subplot = {
  // ========================================
  // 必須メタデータ
  // ========================================

  /** 一意なID */
  id: string;

  /** プロット名 */
  name: string;

  /** プロットタイプ */
  type: PlotType;

  /** 短い概要 */
  summary: string;

  /** プロットのビート構成 */
  beats: PlotBeat[];

  // ========================================
  // キャラクター関連
  // ========================================

  /** このプロットのフォーカスキャラクター */
  focusCharacters: FocusCharacter[];

  /** 関連キャラクター（フォーカス以外） */
  relatedCharacters?: string[];

  // ========================================
  // 構造・関係性
  // ========================================

  /** 構造テンプレートID（外部定義） */
  structureTemplateId?: string;

  /** 親プロットID（サブプロットの場合） */
  parentPlotId?: string;

  /** 子プロットIDリスト */
  childPlotIds?: string[];

  /** このプロットが扱うテーマ */
  themes?: string[];

  // ========================================
  // メタ情報
  // ========================================

  /** 重要度 */
  importance?: PlotImportance;

  /** 表示名のバリエーション */
  displayNames?: string[];

  /** 詳細情報 */
  details?: SubplotDetails;

  /** LSP用の検出ヒント */
  detectionHints?: SubplotDetectionHints;
};

File: internal/domain/subplot.go
package domain

// This file bundles Subplot, PlotBeat, PlotIntersection and SubplotRelations.
// All four types share the same conceptual cluster (a subplot is composed of
// beats and connected to other subplots through intersections), so co-locating
// them keeps the cross-references local and avoids exposing intermediate
// helpers.
//
// Why: Adopt string-typed const enums (e.g. `type SubplotType string`) instead
// of int-iota. Reasons:
//   - JSON round-trip with src/type/v2/subplot.ts uses literal strings as the
//     source of truth; matching the wire representation by value avoids a
//     translation table on every marshal/unmarshal.
//   - Mission spec (WT-A1-5) explicitly mandates this representation.
//   - Other domain bundles (foreshadowing/timeline) may use iota; the
//     entity-type-mapping doc allows divergence per-bundle.
//
// Why: `string | { file: string }` unions in TS map to the shared
// StringOrFileRef helper (`internal/domain/common.go`). Wave-A2-pre で集約済み:
// 旧 inline anonymous struct (Inline/FileRef) は撤去し、Character / Setting /
// Foreshadowing / Timeline と同じ型で統一する。

// SubplotType classifies a plot line by narrative role.
//
// TS source: `"main" | "subplot" | "parallel" | "background"`
// (src/type/v2/subplot.ts).
type SubplotType string

const (
	SubplotTypeMain       SubplotType = "main"
	SubplotTypeSubplot    SubplotType = "subplot"
	SubplotTypeParallel   SubplotType = "parallel"
	SubplotTypeBackground SubplotType = "background"
)

// SubplotStatus tracks lifecycle progression of a subplot.
type SubplotStatus string

const (
	SubplotStatusActive    SubplotStatus = "active"
	SubplotStatusCompleted SubplotStatus = "completed"
)

// SubplotImportance flags how central a subplot is to the overall story.
type SubplotImportance string

const (
	SubplotImportanceMajor SubplotImportance = "major"
	SubplotImportanceMinor SubplotImportance = "minor"
)

// FocusCharacterPriority records how strongly a character drives a subplot.
//
// TS source: `Record<string, "primary" | "secondary">` on Subplot.focusCharacters.
type FocusCharacterPriority string

const (
	FocusCharacterPriorityPrimary   FocusCharacterPriority = "primary"
	FocusCharacterPrioritySecondary FocusCharacterPriority = "secondary"
)

// StructurePosition labels where a beat sits in the narrative arc.
//
// TS source: BeatStructurePosition in src/type/v2/subplot.ts.
// Five-stage arc (setup, rising, climax, falling, resolution).
type StructurePosition string

const (
	StructurePositionSetup      StructurePosition = "setup"
	StructurePositionRising     StructurePosition = "rising"
	StructurePositionClimax     StructurePosition = "climax"
	StructurePositionFalling    StructurePosition = "falling"
	StructurePositionResolution StructurePosition = "resolution"
)

// InfluenceDirection states how influence flows across a PlotIntersection.
//
// TS source: IntersectionInfluenceDirection (forward/backward/mutual).
type InfluenceDirection string

const (
	InfluenceDirectionForward  InfluenceDirection = "forward"
	InfluenceDirectionBackward InfluenceDirection = "backward"
	InfluenceDirectionMutual   InfluenceDirection = "mutual"
)

// InfluenceLevel grades the strength of a PlotIntersection.
type InfluenceLevel string

const (
	InfluenceLevelHigh   InfluenceLevel = "high"
	InfluenceLevelMedium InfluenceLevel = "medium"
	InfluenceLevelLow    InfluenceLevel = "low"
)

// PlotBeat is one narrative beat inside a Subplot. Beats are also the
// connection points referenced by PlotIntersection.
type PlotBeat struct {
	// Required identity & content
	ID                string
	Title             string
	Summary           string
	StructurePosition StructurePosition

	// Optional cross-entity references. Pointers for scalar optionals,
	// nil slices for absent collections (TS `field?: string[]`).
	Chapter             *string
	Characters          []string
	Settings            []string
	TimelineEventID     *string
	PreconditionBeatIDs []string
}

// PlotIntersection links a beat in one subplot to a beat in another, modelling
// how plotlines influence each other.
type PlotIntersection struct {
	// Required identity & link surface.
	ID                 string
	SourceSubplotID    string
	SourceBeatID       string
	TargetSubplotID    string
	TargetBeatID       string
	Summary            string
	InfluenceDirection InfluenceDirection

	// Optional grading.
	InfluenceLevel *InfluenceLevel
}

// SubplotDetails carries optional long-form descriptive fields. Each field
// uses the shared StringOrFileRef union (see file header comment).
type SubplotDetails struct {
	Description StringOrFileRef
	Theme       StringOrFileRef
	Notes       StringOrFileRef
}

// SubplotRelations records cross-entity ID references owned by a Subplot.
// Characters/Settings are required (always present, possibly empty); the rest
// are optional and default to nil.
type SubplotRelations struct {
	Characters      []string
	Settings        []string
	Foreshadowings  []string // optional (nil = absent)
	RelatedSubplots []string // optional
}

// Subplot is a single plot line composed of ordered beats and connected to
// other subplots through intersections.
type Subplot struct {
	// Required identity & metadata
	ID      string
	Name    string
	Type    SubplotType
	Status  SubplotStatus
	Summary string
	Beats   []PlotBeat

	// Optional structural information
	FocusCharacters map[string]FocusCharacterPriority // nil = absent
	Intersections   []PlotIntersection                // nil = absent
	Importance      *SubplotImportance                // nil = absent
	ParentSubplotID *string                           // nil = top-level
	DisplayNames    []string                          // nil = absent
	Details         *SubplotDetails                   // nil = absent
	Relations       *SubplotRelations                 // nil = absent
}
